pkg/kptfile/kptfileutil/util.go

```diff
@@ -261,6 +271,85 @@ func merge(localKf, updatedKf, originalKf *kptfilev1.KptFile) error {
 	return nil
 }
 
+// shouldAddSyntheticMergeName returns true iff all the functions in all the sources
+// doesn't have name field set, it means the user is unaware of name field, and we
+// use image as mergeKey instead of name in such cases
+// it also checks if there are no duplicate image name declarations
+func shouldAddSyntheticMergeName(kfs ...*kptfilev1.KptFile) bool {
+	for _, kf := range kfs {
+		if kf == nil || kf.Pipeline == nil {
+			continue
+		}
+		if !shouldAddSyntheticMergeNameUtil(kf.Pipeline.Mutators) || !shouldAddSyntheticMergeNameUtil(kf.Pipeline.Validators) {
+			return false
+		}
+	}
+	return true
+}
+
+// shouldAddSyntheticMergeNameUtil is util method for shouldAddSyntheticMergeName
+func shouldAddSyntheticMergeNameUtil(fns []kptfilev1.Function) bool {
+	imageNamesSet := sets.String{}
+	for _, fn := range fns {
+		if fn.Name != "" {
+			return false
+		}
+		imageName := strings.Split(fn.Image, ":")[0]
+		if imageNamesSet.Has(imageName) {
+			return false
+		}
+		imageNamesSet.Insert(imageName)
+	}
+	return true
+}
+
+// addNameForMerge adds name field for all the functions if empty
+// name is primarily used as merge-key
+func addNameForMerge(kfs ...*kptfilev1.KptFile) {
+	for _, kf := range kfs {
+		if kf == nil || kf.Pipeline == nil {
+			continue
+		}
+		for i, mutator := range kf.Pipeline.Mutators {
+			kf.Pipeline.Mutators[i] = addName(mutator)
+		}
+		for i, validator := range kf.Pipeline.Validators {
+			kf.Pipeline.Validators[i] = addName(validator)
+		}
+	}
+}
+
+// addName adds name field to the input function if empty
+// name is nothing but image name in this case as we use it as fall back mergeKey
+func addName(fn kptfilev1.Function) kptfilev1.Function {
+	if fn.Name == "" {
```

Comment:
nit: use `if fn.Name != "" {return fn}` to return early and avoid nesting. I think golint should be catching this.

Reply:
Done.